### PR TITLE
02-library

### PR DESCRIPTION
**Pull request by ztcric**

### CODE REVIEW

File: 02-library/cs506/read.py [hunk 1]
```diff
@@ -3,4 +3,14 @@ def read_csv(csv_file_path):
...
         in row major.
     """
-    raise NotImplementedError()
+    res = []
+    with open(csv_file_path, "r") as f:
+        lines = f.readlines()
+        for line in lines:
+            row = []
+            for x in line.split(","):
+                row.append(float(x))
```
Comment (gallettilance) [inline 02-library/cs506/read.py]: what if this is not a float? what if x is an int or a string or a date?

File: 02-library/cs506/sim.py [hunk 2]
```diff
@@ -1,13 +1,38 @@
 def euclidean_dist(x, y):
```
Comment (gallettilance) [inline 02-library/cs506/sim.py]: what if x and y are of different lengths? what if their length is 0?

File: 02-library/cs506/read.py (hunk 1 above)
Comment (ztcric) [reply 02-library/cs506/read.py]: I'm using ast.literal_eval(). Would it solve the issue? I'm a newbie in Python.

File: 02-library/cs506/sim.py (hunk 2 above)
Comment (ztcric) [reply 02-library/cs506/sim.py]: When x and y have different lengths, I get rid of the surplus part and loop through with the shorter length. Not sure if that's correct. Thanks!

Comment (gallettilance) [reply 02-library/cs506/sim.py]: as we discussed in office hours this is incorrect. Please let me know when you have updated these functions :)

Comment (ztcric) [reply 02-library/cs506/sim.py]: Updated. Please take a look. Thanks!

File: 02-library/cs506/read.py (hunk 1 above)
Comment (ztcric) [reply 02-library/cs506/read.py]: Updated. I'm using eval() now. Thank you!

File: 02-library/cs506/read.py [hunk 3]
```diff
@@ -1,6 +1,19 @@
+
 def read_csv(csv_file_path):
     """
-        Given a path to a csv file, return a matrix (list of lists)
-        in row major.
+    Read data from csv file
+    Convert data into its original data type by using ast.literal_eval()
```
Comment (gallettilance) [inline 02-library/cs506/read.py]: comment is incorrect

File: 02-library/cs506/read.py [hunk 4]
```diff
@@ -1,6 +1,19 @@
...
     """
-    raise NotImplementedError()
+    res = []
+    with open(csv_file_path, "r") as f:
+        lines = f.readlines()
+        for line in lines:
+            row = []
+            for x in line.split(","):
+                value = eval(x)
+                row.append(value)
```
Comment (gallettilance) [inline 02-library/cs506/read.py]:
```suggestion
                row.append(eval(x))
```

File: 02-library/cs506/read.py [hunk 5]
```diff
@@ -1,6 +1,19 @@
...
+    return a matrix
     """
-    raise NotImplementedError()
+    res = []
+    with open(csv_file_path, "r") as f:
+        lines = f.readlines()
+        for line in lines:
+            row = []
+            for x in line.split(","):
+                value = eval(x)
```
Comment (gallettilance) [inline 02-library/cs506/read.py]:
```suggestion
```

File: 02-library/cs506/sim.py [hunk 6]
```diff
@@ -1,13 +1,64 @@
...
+    """
+    if len(x) == 0 or len(y) == 0:
+        raise ValueError("lengths must not be zero")
+    x_set = set(x)
+    y_set = set(y)
+    intersection = len(x_set.intersection(y_set))
+    union = len(x_set.union(y_set))
+    sim = intersection/union
+    res = 1 - sim;
+    return res
```
Comment (gallettilance) [inline 02-library/cs506/sim.py]:
```suggestion
    return 1 - intersection/union
```

File: 02-library/cs506/sim.py [hunk 7]
```diff
@@ -1,13 +1,64 @@
...
+    Calculate the jaccard distance between x and y
+    """
+    if len(x) == 0 or len(y) == 0:
+        raise ValueError("lengths must not be zero")
+    x_set = set(x)
+    y_set = set(y)
+    intersection = len(x_set.intersection(y_set))
+    union = len(x_set.union(y_set))
+    sim = intersection/union
+    res = 1 - sim;
```
Comment (gallettilance) [inline 02-library/cs506/sim.py]:
```suggestion
```

File: 02-library/cs506/sim.py [hunk 8]
```diff
@@ -1,13 +1,64 @@
...
+    """
+    Calculate the jaccard distance between x and y
+    """
+    if len(x) == 0 or len(y) == 0:
+        raise ValueError("lengths must not be zero")
+    x_set = set(x)
+    y_set = set(y)
+    intersection = len(x_set.intersection(y_set))
+    union = len(x_set.union(y_set))
+    sim = intersection/union
```
Comment (gallettilance) [inline 02-library/cs506/sim.py]:
```suggestion
```

File: 02-library/cs506/sim.py [hunk 9]
```diff
@@ -1,13 +1,64 @@
...
+    dot = 0
+    x_magnitude = 0
+    y_magnitude = 0
+    for i in range(len(x)):
+        dot += x[i] * y[i]
+        x_magnitude += x[i]**2
+        y_magnitude += y[i]**2
+    x_magnitude = x_magnitude**(1/2)
+    y_magnitude = y_magnitude**(1/2)
+    res = dot / (x_magnitude * y_magnitude)
```
Comment (gallettilance) [inline 02-library/cs506/sim.py]:
```suggestion
    return dot / (x_magnitude * y_magnitude)
```

File: 02-library/cs506/sim.py [hunk 10]
```diff
@@ -1,13 +1,64 @@
...
+    return res
```
Comment (gallettilance) [inline 02-library/cs506/sim.py]:
```suggestion
```

File: 02-library/cs506/read.py (hunk 3 above)
Comment (ztcric) [reply 02-library/cs506/read.py]: Updated. Sorry.

File: 02-library/cs506/sim.py [hunk 11]
```diff
@@ -1,13 +1,61 @@
...
+    intersection = len(x_set.intersection(y_set))
+    union = len(x_set.union(y_set))
+    return 1 - intersection/union
 
 def cosine_sim(x, y):
-    raise NotImplementedError()
-
-# Feel free to add more
+    """
+    Calculate the consine similarity between x and y
```
Comment (gallettilance) [inline 02-library/cs506/sim.py]: Nit: there is a small typo here.

Comment (ztcric) [reply 02-library/cs506/sim.py]: My bad. Fixed. Thanks

**Pull request merged**